packages/instruction-plans/src/__tests__/transaction-planner-base-test.ts

```diff
@@ -0,0 +1,1652 @@
+
+            await expect(planner(messagePackerIx)).resolves.toEqual(
+                sequentialTransactionPlan([
+                    singleTransactionPlan([messagePackerIx.get(0, txPercent(100))]),
+                    singleTransactionPlan([messagePackerIx.get(txPercent(100), txPercent(100))]),
+                    singleTransactionPlan([messagePackerIx.get(txPercent(200), txPercent(50))]),
+                ]),
+            );
+        });
+
+        /**
+         *  [Seq] ───────────────────▶ [Tx: A + B(1, 50%)]
+         *   │
+         *   ├── [A: 50%]
+         *   └── [B(x, 50%)]
+         */
+        it('combines single instruction plans with message packer instruction plans', async () => {
+            expect.assertions(1);
+            const createTransactionMessage = createMockTransactionMessage;
+            const { instruction, singleTransactionPlan, txPercent } = getHelpers(createTransactionMessage);
+            const planner = createBaseTransactionPlanner({ createTransactionMessage });
+
+            const instructionA = instruction('A', txPercent(50));
+            const messagePackerB = createMessagePackerInstructionPlan(txPercent(50));
+
+            await expect(
+                planner(sequentialInstructionPlan([singleInstructionPlan(instructionA), messagePackerB])),
+            ).resolves.toEqual(singleTransactionPlan([instructionA, messagePackerB.get(0, txPercent(50))]));
+        });
+
+        /**
+         *  [Par] ────────────────────▶ [Par]
+         *   │                           │
+         *   └── [A(x, 250%)]            ├── [Tx: A(1, 100%)]
+         *                               ├── [Tx: A(2, 100%)]
+         *                               └── [Tx: A(3, 50%)]
+         */
+        it('can handle parallel message packer instruction plans', async () => {
+            expect.assertions(1);
+            const createTransactionMessage = createMockTransactionMessage;
+            const { singleTransactionPlan, txPercent } = getHelpers(createTransactionMessage);
+            const planner = createBaseTransactionPlanner({ createTransactionMessage });
+            const messagePackerA = createMessagePackerInstructionPlan(txPercent(250));
+
+            await expect(planner(parallelInstructionPlan([messagePackerA]))).resolves.toEqual(
+                parallelTransactionPlan([
+                    singleTransactionPlan([messagePackerA.get(0, txPercent(100))]),
+                    singleTransactionPlan([messagePackerA.get(txPercent(100), txPercent(100))]),
+                    singleTransactionPlan([messagePackerA.get(txPercent(200), txPercent(50))]),
+                ]),
+            );
+        });
+
+        /**
+         *  [NonDivSeq] ──────────────▶ [NonDivSeq]
+         *   │                           │
+         *   └── [A(x, 250%)]            ├── [Tx: A(1, 100%)]
+         *                               ├── [Tx: A(2, 100%)]
+         *                               └── [Tx: A(3, 50%)]
+         */
```

Comment:
What does it mean for something to be non-divisible-sequential at the transaction level like this? Don't the transaction boundaries themselves imply that the plan has been divided? Does this imply they _must_ be landed in a bundle.

Reply:
That's right, it implies they must be using a `TransactionPlanExecutor` that is able to tackle that wider atomicity. By having a non-divisible transaction plan, we're able to push that decision to the `TransactionPlanExecutor` instead of failing when transaction bundles could be used.